FEXCore/Source/Interface/Core/OpcodeDispatcher/X87.cpp

```diff
@@ -72,757 +55,313 @@ void OpDispatchBuilder::SetX87FTW(Ref FTW) {
-  auto top = GetX87Top();
-  auto a = _LoadContextIndexed(top, 16, MMBaseOffset(), 16, FPRClass);
-
-  DeriveOp(result, IROp, _F80Round(a));
-
-  if (IROp == IR::OP_F80SIN || IROp == IR::OP_F80COS) {
-    // TODO: ACCURACY: should check source is in range –2^63 to +2^63
-    SetRFLAG<FEXCore::X86State::X87FLAG_C2_LOC>(_Constant(0));
-  }
-
-  // Write to ST[TOP]
-  _StoreContextIndexed(result, top, 16, MMBaseOffset(), 16, FPRClass);
-}
-
-void OpDispatchBuilder::X87BinaryOp(OpcodeArgs, FEXCore::IR::IROps IROp) {
-  auto top = GetX87Top();
-
-  auto mask = _Constant(7);
-  Ref st1 = _And(OpSize::i32Bit, _Add(OpSize::i32Bit, top, _Constant(1)), mask);
-
-  auto a = _LoadContextIndexed(top, 16, MMBaseOffset(), 16, FPRClass);
-  st1 = _LoadContextIndexed(st1, 16, MMBaseOffset(), 16, FPRClass);
-
-  DeriveOp(result, IROp, _F80Add(a, st1));
-
-  if (IROp == IR::OP_F80FPREM || IROp == IR::OP_F80FPREM1) {
-    // TODO: Set C0 to Q2, C3 to Q1, C1 to Q0
-    SetRFLAG<FEXCore::X86State::X87FLAG_C2_LOC>(_Constant(0));
   }
 
-  // Write to ST[TOP]
-  _StoreContextIndexed(result, top, 16, MMBaseOffset(), 16, FPRClass);
-}
+  // top of stack is at offset zero
+  _F80MulValue(0, arg);
 
-void OpDispatchBuilder::X87ModifySTP(OpcodeArgs, bool Inc) {
-  auto orig_top = GetX87Top();
-  if (Inc) {
-    auto top = _And(OpSize::i32Bit, _Add(OpSize::i32Bit, orig_top, _Constant(1)), _Constant(7));
-    SetX87Top(top);
-  } else {
-    auto top = _And(OpSize::i32Bit, _Sub(OpSize::i32Bit, orig_top, _Constant(1)), _Constant(7));
-    SetX87Top(top);
+  if (Op->TableInfo->Flags & X86Tables::InstFlags::FLAGS_POP) {
+    _PopStackDestroy();
   }
 }
 
-void OpDispatchBuilder::X87SinCos(OpcodeArgs) {
-  auto orig_top = GetX87Top();
-  auto top = _And(OpSize::i32Bit, _Sub(OpSize::i32Bit, orig_top, _Constant(1)), _Constant(7));
-  SetX87ValidTag(top, true);
-  SetX87Top(top);
-
-  auto a = _LoadContextIndexed(orig_top, 16, MMBaseOffset(), 16, FPRClass);
-
-  auto sin = _F80SIN(a);
-  auto cos = _F80COS(a);
-
-  // TODO: ACCURACY: should check source is in range –2^63 to +2^63
-  SetRFLAG<FEXCore::X86State::X87FLAG_C2_LOC>(_Constant(0));
-
-  // Write to ST[TOP]
-  _StoreContextIndexed(sin, orig_top, 16, MMBaseOffset(), 16, FPRClass);
-  _StoreContextIndexed(cos, top, 16, MMBaseOffset(), 16, FPRClass);
-}
+void OpDispatchBuilder::FDIV(OpcodeArgs, size_t Width, bool Integer, bool Reverse, OpDispatchBuilder::OpResult ResInST0) {
+  CurrentHeader->HasX87 = true;
 
-void OpDispatchBuilder::X87FYL2X(OpcodeArgs) {
-  bool Plus1 = Op->OP == 0x01F9; // FYL2XP
+  if (Op->Src[0].IsNone()) {
+    const auto offset = Op->OP & 7;
+    const auto st0 = 0;
 
-  auto orig_top = GetX87Top();
-  // if we are popping then we must first mark this location as empty
-  SetX87ValidTag(orig_top, false);
-  auto top = _And(OpSize::i32Bit, _Add(OpSize::i32Bit, orig_top, _Constant(1)), _Constant(7));
-  SetX87Top(top);
+    if (Reverse) {
```

Comment:
can combine the conditions: `if (Reverse ^ (ResInST0 == OpResult::RES_STI))`

and then drop the latter half

Reply:
I guess you'd still need a `Ref Result = (ResInST0 == OpR....) ? offset : st0;` or something. But given my eyes glazed over that difference the first time, that'd probably be more readable anyhow.

Reply:
Nice catch!